lib/actions/appleMusic.js

```diff
@@ -38,8 +52,127 @@ function appleMusic(player, values) {
   
   var uri = uriTemplates[type].format({id: encodeURIComponent(trackID)});
 
+  // Parse the iTunes search results and load queue with tracks.  The search name can be an artist name OR a song title  
+  var handleiTunesResponse = function (error, tracksJson) {
+    if (!error) {
+      try { 
+        tracksJson = JSON.parse(tracksJson);
+        console.log("Found " + tracksJson.resultCount + " tracks for " + trackID);
+        if (tracksJson.resultCount > 0) {
+        
+          // Find the first streamable track
+          var trkNum; 
+          for (trkNum=0;(trkNum < tracksJson.resultCount) && !tracksJson.results[trkNum].isStreamable ;trkNum++) {
+          }
+
+          // There is at least one streamable track  
+          if (trkNum < tracksJson.resultCount) {
+            var prevArtist = tracksJson.results[trkNum].artistName;
+            var prevTrack = tracksJson.results[trkNum].trackName;
+            var artistCount = 1;
+            var trackCount = 1;
+            
+            // Determine if the search results are many tracks for an artist OR a specific song title.
+            for (var i=trkNum; i < tracksJson.resultCount;i++) {
+              var track = tracksJson.results[i];
+              
+              if (track.artistName != prevArtist) {
+                artistCount++;
+                prevArtist = track.artistName;
+              }
+              if (track.trackName != prevTrack) {
+                trackCount++;
+                prevTrack = track.trackName;
+              }  
+            }
+          
+            // If there are more track differences that artist differences, then load all of the artist's tracks into the queue, ELSE play just the first track now (iTunes returns the most popular track first) 
+            if (trackCount > artistCount) {
+              player.coordinator.removeAllTracksFromQueue();
+              setTimeout(function() {
+                // Add the first track and start it playing
+                addTrackToQueue(tracksJson.results[trkNum], true);
+                
+                // Add each of the remaining tracks after a delay interval to keep from pounding Sonos
+                var intervalID = setInterval(function() {
+                  // Check see if this track is a duplicate of a track that has already been queued and skip it if so
+                  var found = false;
```

Comment:
It seems that the problem is that addURIToQueue is not acting as I would expect related to the callback.  Looking at the function in player.js, it looks like a passed callback isn't being called until later but the function is continuing anyway??  All of the "callback 3" logs come AFTER all of the tracks are added to the queue.  I would expect to see them called after each track is added.  Odd, and defeats the whole purpose of trying to wait on addURIToQueue.  Any suggestions? Code:
                async.each(tracksJson.results,
                  function(track,callback) {
                    trkNum++;
console.log("trkNum=" + trkNum);                    
                    // Check see if this track is a duplicate of a track that has already been queued and skip it if so
                    if (track.isStreamable) {
                      var found = false;

                      for (var j=0; (j < trkNum-1) && !found ; j++) {
                        dupe_check = tracksJson.results[j];
                        found = (track.trackName == dupe_check.trackName);
                      }

                      if (!found) {
                        var metadataID = METADATA_URI_STARTERS[type] + encodeURIComponent('song:' + track.trackId);     

                        var metadata = metadataTemplate.format({
                          metadataID: metadataID,
                          parentURI: PARENTS[type],
                          clazz: CLASSES[type]
                        });
            
                        var uri = uriTemplates[type].format({id: encodeURIComponent('song:' + track.trackId)});

                        if (first) {
                          var nextTrackNo = player.coordinator.state.trackNo + 1;
                          player.coordinator.addURIToQueue(uri, metadata, true, nextTrackNo, function (error, res) {
                            if (!error) {
                              //need this to tell sonos to use queue (it may be playing from line in, etc)
                              var queueURI = 'x-rincon-queue:' + player.uuid + '#0';
                              player.setAVTransportURI(queueURI, '', function (callback) {
                                if (nextTrackNo > 1) {
                                  player.coordinator.nextTrack(function (error) {
                                    player.coordinator.play(function(error) {
                                      callback();
                                    });
                                  });
                                } else {
                                  player.coordinator.play(function(error) {
console.log("cb 1");                    
                                    callback();
                                  });
                                }
                              });
                            } else {
console.log("cb 2");                    
                              callback(error);
                            }
                          });
                          first = false;
                        } else {
                          var nextTrackNo = player.coordinator.state.trackNo + 1;
                          player.coordinator.addURIToQueue(uri, metadata, function (error, res) {
console.log("cb 3 trkNum=" + trkNum);                    
                            callback();
                          });
                        }
                      } else {
console.log("cb 4");                    
                        callback();
                      }
                    } else {
console.log("cb 5");                    
                      callback();
                    }
                  },
                  function() {
console.log("finished");                    
                  });